File: app-frontend/src/types/documents.ts
export const DocumentStatus = {
  UPLOADED: "uploaded",
  PROCESSING: "processing",
  EXTRACTED: "extracted",
  SUMMARIZED: "summarized",
  FAILED: "failed"
} as const;

export type DocumentStatus = typeof DocumentStatus[keyof typeof DocumentStatus];


export interface DocumentResponse {
    document_id: string;
    user_id: string;
    file_name: string;
    mime_type: string;
    file_size_bytes: number;
    status: DocumentStatus;
    uploaded_at: string;
    error_message?: string | null;
}


export interface DocumentListResponse {
    documents: DocumentResponse [];
    total: number;
}


export interface UploadResponse {
    document_id: string;
    status: DocumentStatus;
    message: string;
}

export interface SummaryResponse {
    summary_id: string;
    document_id: string;
    summary_text: string;
    reading_level_target: string;
    created_at: string;
    disclaimer: string;

}


export interface ChatRequest {
    question: string;
}


export interface ChatResponse {
    message_id: string;
    document_id: string;
    question: string;
    answer: string;
    created_at: string;
    disclaimer: string;
}

export interface FeedbackRequest {
    rating: number;
}




export const UnderstandingRating = {
    YES: "yes",
    SOMEWHAT: "somewhat",
    NO: "no"
} as const;

export type UnderstandingRating = keyof typeof UnderstandingRating;


export interface UnderstandingRequest {
    rating: string;
}


export interface PrepResponse {
    prep_id: string;
    document_id: string;
    questions: string[];
    created_at: string;
    notes: string;
}



export interface UnderstadingResponse {
    id: string;
    summary_id: string;
    rating: string;
    created_at: string;
}


export interface KpiReadingLevel {
    total_summaries: number;
    avg_reading_level?: number;
    at_or_below_grade_6: number;
    pct_on_target?: number;
}

export interface KpiQuality {
    total_summaries: number;
    passed: number;
    failed: number;
    pass_rate_pct?: number;
}

export interface KpiSatisfaction {
    rated_messages: number;
    avg_rating?: number;
    positive: number;
    negative: number;
}


export interface DashboardResponse {
    documents: DocumentResponse[];
    total_documents: number;
    total_summaries: number;
    total_questions_asked: number;
    avg_seconds_to_summary?: number;
    reading_level?: KpiReadingLevel;
    quality?: KpiQuality;
    satisfaction?: KpiSatisfaction;
}


export const MedBridgeErrorCode = {
    groqTimeout : "GROQ_TIMEOUT",
    groqUnavailable : "GROQ_UNAVAILABLE",
     docNotFound : "DOC_NOT_FOUND",
     extractionFailed : "EXTRACTION_FAILED",
     unauthorized : "UNAUTHORIZED"
} as const;

export type MedBridgeErrorCode = typeof MedBridgeErrorCode[keyof typeof MedBridgeErrorCode];

export interface ErrorEnvelope {
    success: boolean;
    error_code: MedBridgeErrorCode;
    message: string;
    retry_after?: number;
};






//! Duplicate interface in feature file:

export interface ReminderCreate {
    title: string;
    due_date?: string;
    notes?: string;
}

export interface ReminderResponse {
    reminder_id: string;
    user_id: string;
    document_id?: string;
    title: string;
    due_date?: string;
    notes?: string;
    created_at: string;
}

File: app-backend/app/schemas/documents.py
from pydantic import BaseModel, Field
from typing import Optional
from datetime import datetime
from enum import Enum
from app.schemas.medications import MedicationCreate, MedicationResponse
from app.schemas.patient_profile import PatientProfileResponse
from app.utils.validation import NonEmptyStr


class DocumentStatus(str, Enum):
    UPLOADED = "uploaded"
    PROCESSING = "processing"
    EXTRACTED = "extracted"
    SUMMARIZED = "summarized"
    READY = "ready"
    FAILED = "failed"


class PendingMedication(BaseModel):
    name: str
    dose: Optional[str] = None
    dosage: Optional[str] = None
    frequency: Optional[str] = None
    route: Optional[str] = None
    prescribing_provider: Optional[str] = None
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    reason: Optional[str] = None
    notes: Optional[str] = None


class ExtractedPrescriptionData(BaseModel):
    name: str
    dosage: Optional[str] = None
    frequency: Optional[str] = None
    route: Optional[str] = None
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    prescribing_provider: Optional[str] = None
    reason: Optional[str] = None
    notes: Optional[str] = None


class DocumentResponse(BaseModel):
    document_id: str
    user_id: str
    file_name: str
    mime_type: str
    file_size_bytes: int
    status: DocumentStatus
    uploaded_at: datetime
    error_message: Optional[str] = None
    extracted_text: Optional[str] = None
    is_prescription: bool = False
    pending_medications: Optional[list[PendingMedication]] = None


class ConfirmPrescriptionMedicationsRequest(BaseModel):
    medications: list[MedicationCreate] = Field(..., min_length=1)


class ConfirmPrescriptionMedicationsResponse(BaseModel):
    document_id: str
    medications: list[MedicationResponse]


class DocumentListResponse(BaseModel):
    documents: list[DocumentResponse]
    total: int


class UploadResponse(BaseModel):
    document_id: str
    file_name: str
    is_prescription: bool = False
    extracted_data: Optional[ExtractedPrescriptionData] = None
    status: DocumentStatus
    message: str = "Document uploaded successfully. Processing has started."


class SummaryResponse(BaseModel):
    summary_id: str
    document_id: str
    summary_text: str
    reading_level_target: str = "4th grade"
    created_at: datetime
    disclaimer: str = (
        "MedBridge explains information from your uploaded document in plain language. "
        "It does not provide medical advice, diagnosis, or treatment recommendations. "
        "Please contact a healthcare provider for medical decisions."
    )


class ChatRequest(BaseModel):
    question: NonEmptyStr = Field(..., min_length=1, max_length=2000)


class ChatResponse(BaseModel):
    message_id: str
    document_id: str
    question: str
    answer: str
    created_at: datetime
    disclaimer: str = (
        "This answer is based only on your uploaded document. "
        "MedBridge does not provide medical advice. "
        "Please consult a healthcare provider for medical decisions."
    )


class FeedbackRequest(BaseModel):
    rating: int = Field(..., ge=1, le=5)


class ReminderCreate(BaseModel):
    title: NonEmptyStr = Field(..., min_length=1, max_length=255)
    due_date: Optional[datetime] = None
    notes: Optional[str] = None


class ReminderResponse(BaseModel):
    reminder_id: str
    user_id: str
    document_id: Optional[str] = None
    title: str
    due_date: Optional[datetime]
    notes: Optional[str]
    created_at: datetime


# ── Understanding rating ──────────────────────────────────────────────────────

class UnderstandingRating(str):
    YES = "yes"
    SOMEWHAT = "somewhat"
    NO = "no"


class UnderstandingRequest(BaseModel):
    rating: str = Field(..., pattern="^(yes|somewhat|no)$")


class UnderstandingResponse(BaseModel):
    id: str
    summary_id: str
    rating: str
    created_at: datetime


# ── Appointment prep ──────────────────────────────────────────────────────────

class PrepResponse(BaseModel):
    prep_id: str
    document_id: str
    questions: list[str]
    created_at: datetime
    note: str = (
        "These questions are generated from your document to help you "
        "prepare for a conversation with your healthcare provider. "
        "They are not medical advice."
    )


# ── Dashboard (extended) ──────────────────────────────────────────────────────

class KpiReadingLevel(BaseModel):
    total_summaries: int = 0
    avg_reading_level: Optional[float] = None
    at_or_below_grade_6: int = 0
    pct_on_target: Optional[float] = None

class KpiQuality(BaseModel):
    total_summaries: int = 0
    passed: int = 0
    failed: int = 0
    pass_rate_pct: Optional[float] = None

class KpiSatisfaction(BaseModel):
    rated_messages: int = 0
    avg_rating: Optional[float] = None
    positive: int = 0
    negative: int = 0

class DashboardResponse(BaseModel):
    user: PatientProfileResponse
    profile_complete: bool
    documents: list[DocumentResponse]
    total_documents: int
    total_summaries: int
    total_questions_asked: int
    avg_seconds_to_summary: Optional[float] = None
    reading_level: Optional[KpiReadingLevel] = None
    quality: Optional[KpiQuality] = None
    satisfaction: Optional[KpiSatisfaction] = None
